### PR TITLE
ETCD status enhancements

### LINKED ISSUE #168: [Feature] Enhance the Etcd resource status structure as proposed in the multi-node proposal
**Feature (What you would like to be added):**

Enhance the Etcd resource status structure according to the changes proposed [here](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/multi-node/README.md#status) while maintaining backward compatibility for the consumers of `Etcd` resource status (such as the `gardenlet`).

**Motivation (Why is this needed?):**

Pick individually executable pieces of the [multi-node proposal](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/multi-node/README.md).

**Approach/Hint to the implement solution (optional):**

For backward compatibility, the existing status fields and the values in them need to maintained as they are. In both the main etcd-druid controller (especially, [here](https://github.com/gardener/etcd-druid/blob/e7f635d6d55fbc93380a3c1b9d1821705c619978/controllers/etcd_controller.go#L1140-L1175) and [here](https://github.com/gardener/etcd-druid/blob/e7f635d6d55fbc93380a3c1b9d1821705c619978/controllers/etcd_controller.go#L1245-L1256)) as well as the newly separated [custodian controller](https://github.com/gardener/etcd-druid/pull/133).

### PR DESCRIPTION
**Pull request by timuthy**
**How to categorize this PR?**

/area scalibility
/kind api-change
/priority normal

**What this PR does / why we need it**:
This PR enhances the `etcd` status sub-resource by a `status.members` section. As opposed to the design, I decided to rename `lastHeartbeatTime` -> `lastUpdateTime` to be consistent with `status.conditions[].lastUpdateTime` (/cc @amshuman-kr).

**Which issue(s) this PR fixes**:
Fixes #168

**Special notes for your reviewer**:
Along the way this PR:
- Removes some unused constants and structs (status related only).
- Improves API documentation of several fields.
- Updates the `etcd` resource sample.
- Fixes `make generate`.

**Release note**:

```other operator
NONE.
```

### CODE REVIEW

File: api/v1alpha1/etcd_types.go [hunk 1]
```diff
@@ -239,103 +237,116 @@ const (
...
+	// ID is the ID of the etcd member.
+	ID string `json:"id"`
+	// Role is the role in the etcd cluster, either `Member` or `Learner`.
+	Role EtcdRole `json:"role"`
+	// Status of the condition, one of True, False, Unknown.
+	Status EtcdMemberConditionStatus `json:"status"`
+	// The reason for the condition's last transition.
+	Reason string `json:"reason"`
+	// LastUpdateTime is the last time this condition was updated.
+	LastUpdateTime metav1.Time `json:"lastHeartbeatTime"`
```
Comment (amshuman-kr) [inline api/v1alpha1/etcd_types.go:307]:
```suggestion
	LastUpdateTime metav1.Time `json:"lastUpdateTime"`
```

File: api/v1alpha1/zz_generated.deepcopy.go [hunk 2]
```diff
@@ -19,7 +19,7 @@
 package v1alpha1
 
 import (
-	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/core/v1"
```
Comment (amshuman-kr) [inline api/v1alpha1/zz_generated.deepcopy.go:22]:
```suggestion
	v1 "k8s.io/api/core/v1"
```

`make check` is failing without the import alias.

**Review by amshuman-kr — CHANGES_REQUESTED** [review]
Thanks a lot for the PR @timuthy! Also, for cleaning up some of the unused fields.

Just a couple of minor comments. Otherwise, looks good to merge.

**Review by amshuman-kr — APPROVED** [review]
Thanks for the changes! LGTM

**Comment by amshuman-kr** [comment]
The following test case failed in the pipeline. But `make test` succeeded for me locally. Not sure if it is just a flaky test.

@abdasgupta Can you please try `make test` as well? If it works then we can merge it.

**Comment by amshuman-kr** [comment]
@abdasgupta Also, can you please try rebasing your changes to use `Status().Patch()` to fix #171 on top of this PR to see if the `DeepCopy` now works?

**Review by amshuman-kr — CHANGES_REQUESTED** [review]
@timuthy After separating `api` as a go module, any changes to api needs to be revendored. This is the [same problem](https://github.com/gardener/etcd-druid/issues/171#issuecomment-844046045) I found in #171 . Could you please `make revendor`?

**Comment by timuthy** [comment]
Thanks for the hint @amshuman-kr, done as requested in 9e51d34032db7c6e92320a7518dc428dbd31d25a.

**Comment by timuthy** [comment]
I needed to update the Gardener dependency as well (because of https://github.com/gardener/gardener/pull/3996) but then realized that Go 1.16 is additionally required (see failing `check` step). This needs to be done in a separate PR because the pipeline does not respect the changes in this PR.

**Comment by timuthy** [comment]
I'll rebase one https://github.com/gardener/etcd-druid/pull/175 is merged (/cc @amshuman-kr)

**Comment by amshuman-kr** [comment]
@timuthy Thanks for the changes and for #175. I want to cut a hot-fix release `v0.5.1` with "Removed locks before updating ETCD status. #174" to avoid potentially encountering performance regression in canary landscapes. Is it ok if I merge #175 and #173 after #174 is merged (because #173 and #174 have conflict in the generated `DeepCopy` functions? 

You may have to resolve conflict while rebasing after #174 is merged (or simply `make generate; make revendor`). I am sorry for the inconvenience.

File: Makefile [hunk 3]
```diff
@@ -65,8 +65,8 @@ deploy: manifests
 # Generate manifests e.g. CRD, RBAC etc.
 .PHONY: manifests
```
Comment (abdasgupta) [inline Makefile:66]: Would you mind fixing `make revendor` as well? I see following error:
```
github.com/gardener/etcd-druid/hack imports
        github.com/gardener/gardener/.github: malformed import path "github.com/gardener/gardener/.github": leading dot in path element
github.com/gardener/etcd-druid/hack imports
        github.com/gardener/gardener/hack/.ci: malformed import path "github.com/gardener/gardener/hack/.ci": leading dot in path element
make: *** [Makefile:31: revendor] Error 1
```

**Comment by amshuman-kr** [comment]
> I'll rebase once #175 is merged (/cc @amshuman-kr)

@timuthy I have merged #174 and #175. Could you please rebase?

File: Makefile (hunk 3 above)
Comment (timuthy) [reply Makefile:66]: Which Golang version do you use? I think this was a bug in Go 1.16.{0,1} (see this issue https://github.com/golang/go/issues/43985).

**Comment by timuthy** [comment]
> > I'll rebase once #175 is merged (/cc @amshuman-kr)
> 
> @timuthy I have merged #174 and #175. Could you please rebase?

Done. PTAL.

**Review by amshuman-kr — APPROVED** [review]
Thanks for rebasing. LGTM

File: Makefile (hunk 3 above)
Comment (abdasgupta) [reply Makefile:66]: I see. Thanks.

**Pull request merged**